File: packages/go/builders.go
package alexsdk

import (
	"encoding/json"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strings"
)

// PackOptions carries advanced options for builder pack invocations.
type PackOptions struct {
	SrcDir string
}

// PackOpt mutates PackOptions.
type PackOpt func(*PackOptions)

// WithSrcDir packs from an existing source dir rather than materialising a
// temp dir of staged files.
func WithSrcDir(dir string) PackOpt {
	return func(o *PackOptions) { o.SrcDir = dir }
}

type stagedFile struct {
	archivePath string
	srcAbs      string
}

type base struct {
	manifest Manifest
	staged   []stagedFile
}

func newBase(name, version string, kind Kind, config any) base {
	raw, _ := json.Marshal(config) // typed config; cannot fail
	return base{
		manifest: Manifest{
			SchemaVersion: "2",
			Name:          name,
			Version:       version,
			Kind:          kind,
			Description:   "",
			Config:        raw,
		},
	}
}

func (b *base) setConfig(c any) {
	raw, _ := json.Marshal(c)
	b.manifest.Config = raw
}

func (b *base) ensurePerms() *Permissions {
	if b.manifest.Permissions == nil {
		b.manifest.Permissions = &Permissions{}
	}
	return b.manifest.Permissions
}

func (b *base) build() (*Manifest, error) {
	// Deep copy via JSON round-trip so callers can keep mutating the builder.
	raw, err := json.Marshal(&b.manifest)
	if err != nil {
		return nil, fmt.Errorf("marshal manifest: %w", err)
	}
	var copy Manifest
	if err := json.Unmarshal(raw, &copy); err != nil {
		return nil, fmt.Errorf("clone manifest: %w", err)
	}
	if err := AssertValid(&copy); err != nil {
		return nil, err
	}
	return &copy, nil
}

func (b *base) packInternal(outPath string, opts []PackOpt) (*Manifest, error) {
	o := PackOptions{}
	for _, f := range opts {
		f(&o)
	}
	manifest, err := b.build()
	if err != nil {
		return nil, err
	}
	if o.SrcDir != "" {
		if err := writeManifestFile(o.SrcDir, manifest); err != nil {
			return nil, err
		}
		return Pack(o.SrcDir, outPath)
	}
	dir, err := os.MkdirTemp("", "alex-sdk-")
	if err != nil {
		return nil, fmt.Errorf("mkdtemp: %w", err)
	}
	if err := writeManifestFile(dir, manifest); err != nil {
		return nil, err
	}
	for _, s := range b.staged {
		dest := filepath.Join(dir, s.archivePath)
		if err := os.MkdirAll(filepath.Dir(dest), 0o755); err != nil {
			return nil, fmt.Errorf("mkdir %s: %w", filepath.Dir(dest), err)
		}
		if err := copyFile(s.srcAbs, dest); err != nil {
			return nil, fmt.Errorf("stage %s: %w", s.archivePath, err)
		}
	}
	return Pack(dir, outPath)
}

func writeManifestFile(dir string, m *Manifest) error {
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return fmt.Errorf("mkdir %s: %w", dir, err)
	}
	data, err := json.MarshalIndent(m, "", "  ")
	if err != nil {
		return fmt.Errorf("encode manifest: %w", err)
	}
	data = append(data, '\n')
	path := filepath.Join(dir, "atool.json")
	if err := os.WriteFile(path, data, 0o644); err != nil {
		return fmt.Errorf("write %s: %w", path, err)
	}
	return nil
}

func copyFile(src, dst string) error {
	in, err := os.Open(src)
	if err != nil {
		return err
	}
	defer in.Close()
	out, err := os.Create(dst)
	if err != nil {
		return err
	}
	defer out.Close()
	_, err = io.Copy(out, in)
	return err
}

// Tool is a fluent builder for kind=tool packages.
type Tool struct {
	base
	cfg ToolConfig
}

// NewTool constructs a Tool builder with the given name and version.
func NewTool(name, version string) *Tool {
	t := &Tool{cfg: ToolConfig{Kind: "tool"}}
	t.base = newBase(name, version, KindTool, t.cfg)
	return t
}

func (t *Tool) sync() *Tool { t.setConfig(t.cfg); return t }

// Description sets the human-readable description.
func (t *Tool) Description(d string) *Tool { t.manifest.Description = d; return t }

// Author sets the author field.
func (t *Tool) Author(a string) *Tool { t.manifest.Author = a; return t }

// License sets the license field.
func (t *Tool) License(l string) *Tool { t.manifest.License = l; return t }

// RequiresAlexandria sets the minimum Alexandria version.
func (t *Tool) RequiresAlexandria(v string) *Tool { t.manifest.RequiresAlexandria = v; return t }

// Dependency appends a dependency to the manifest.
func (t *Tool) Dependency(d Dependency) *Tool {
	t.manifest.Dependencies = append(t.manifest.Dependencies, d)
	return t
}

// Dependencies replaces the dependencies slice.
func (t *Tool) Dependencies(ds []Dependency) *Tool { t.manifest.Dependencies = ds; return t }

// File appends a single file entry.
func (t *Tool) File(f FileEntry) *Tool { t.manifest.Files = append(t.manifest.Files, f); return t }

// Files replaces the files slice.
func (t *Tool) Files(fs []FileEntry) *Tool { t.manifest.Files = fs; return t }

// StageFile stages a file from disk to be copied into the temp src dir at
// pack time, and appends a matching files[] entry.
func (t *Tool) StageFile(srcPath string, entry FileEntry) *Tool {
	abs, _ := filepath.Abs(srcPath)
	t.staged = append(t.staged, stagedFile{archivePath: entry.ArchivePath, srcAbs: abs})
	return t.File(entry)
}

// ProvidesTools sets the permissions.provides_tools list.
func (t *Tool) ProvidesTools(s []string) *Tool { t.ensurePerms().ProvidesTools = s; return t }

// NeedsTools sets the permissions.needs_tools list.
func (t *Tool) NeedsTools(s []string) *Tool { t.ensurePerms().NeedsTools = s; return t }

// SuggestedRole sets the permissions.suggested_role hint.
func (t *Tool) SuggestedRole(r string) *Tool { t.ensurePerms().SuggestedRole = r; return t }

// Binary sets config.binary.
func (t *Tool) Binary(p string) *Tool { t.cfg.Binary = p; return t.sync() }

// Port sets config.default_port.
func (t *Tool) Port(p int) *Tool { v := p; t.cfg.DefaultPort = &v; return t.sync() }

// Transport sets config.transport (http|sse).
func (t *Tool) Transport(s string) *Tool { t.cfg.Transport = s; return t.sync() }

// Args sets config.args.
func (t *Tool) Args(a []string) *Tool { t.cfg.Args = a; return t.sync() }

// K8sImage sets config.k8s_image.
func (t *Tool) K8sImage(img string) *Tool { t.cfg.K8sImage = img; return t.sync() }

// K8sCapabilities sets config.k8s_capabilities.
func (t *Tool) K8sCapabilities(c []string) *Tool { t.cfg.K8sCapabilities = c; return t.sync() }

// K8sPort sets config.k8s_port.
func (t *Tool) K8sPort(p int) *Tool { v := p; t.cfg.K8sPort = &v; return t.sync() }

// K8sTransport sets config.k8s_transport (grpc|http|sse).
func (t *Tool) K8sTransport(s string) *Tool { t.cfg.K8sTransport = s; return t.sync() }

// K8sResources sets config.k8s_resources.
func (t *Tool) K8sResources(r K8sResources) *Tool { t.cfg.K8sResources = &r; return t.sync() }

// K8sMinWarm sets config.k8s_min_warm.
func (t *Tool) K8sMinWarm(n int) *Tool { v := n; t.cfg.K8sMinWarm = &v; return t.sync() }

// K8sIdleTimeout sets config.k8s_idle_timeout_seconds.
func (t *Tool) K8sIdleTimeout(seconds int) *Tool {
	v := seconds
	t.cfg.K8sIdleTimeoutSeconds = &v
	return t.sync()
}

// Build returns the validated manifest.
func (t *Tool) Build() (*Manifest, error) { t.sync(); return t.base.build() }

// Pack writes a .atool to outPath. See PackOpt for options.
func (t *Tool) Pack(outPath string, opts ...PackOpt) (*Manifest, error) {
	t.sync()
	return t.base.packInternal(outPath, opts)
}

// Agent is a fluent builder for kind=agent packages.
type Agent struct {
	base
	cfg AgentConfig
}

// NewAgent constructs an Agent builder.
func NewAgent(name, version string) *Agent {
	a := &Agent{cfg: AgentConfig{Kind: "agent"}}
	a.base = newBase(name, version, KindAgent, a.cfg)
	return a
}

func (a *Agent) sync() *Agent { a.setConfig(a.cfg); return a }

// Description sets the human-readable description.
func (a *Agent) Description(d string) *Agent { a.manifest.Description = d; return a }

// Author sets the author field.
func (a *Agent) Author(s string) *Agent { a.manifest.Author = s; return a }

// License sets the license field.
func (a *Agent) License(l string) *Agent { a.manifest.License = l; return a }

// RequiresAlexandria sets the minimum Alexandria version.
func (a *Agent) RequiresAlexandria(v string) *Agent { a.manifest.RequiresAlexandria = v; return a }

// Dependency appends a dependency.
func (a *Agent) Dependency(d Dependency) *Agent {
	a.manifest.Dependencies = append(a.manifest.Dependencies, d)
	return a
}

// Dependencies replaces the dependencies slice.
func (a *Agent) Dependencies(ds []Dependency) *Agent { a.manifest.Dependencies = ds; return a }

// File appends a single file entry.
func (a *Agent) File(f FileEntry) *Agent { a.manifest.Files = append(a.manifest.Files, f); return a }

// Files replaces the files slice.
func (a *Agent) Files(fs []FileEntry) *Agent { a.manifest.Files = fs; return a }

// StageFile stages a file from disk and appends a matching files[] entry.
func (a *Agent) StageFile(srcPath string, entry FileEntry) *Agent {
	abs, _ := filepath.Abs(srcPath)
	a.staged = append(a.staged, stagedFile{archivePath: entry.ArchivePath, srcAbs: abs})
	return a.File(entry)
}

// ProvidesTools sets permissions.provides_tools.
func (a *Agent) ProvidesTools(s []string) *Agent { a.ensurePerms().ProvidesTools = s; return a }

// NeedsTools sets permissions.needs_tools.
func (a *Agent) NeedsTools(s []string) *Agent { a.ensurePerms().NeedsTools = s; return a }

// SuggestedRole sets permissions.suggested_role.
func (a *Agent) SuggestedRole(r string) *Agent { a.ensurePerms().SuggestedRole = r; return a }

// SystemPrompt sets config.system_prompt.
func (a *Agent) SystemPrompt(s string) *Agent { a.cfg.SystemPrompt = s; return a.sync() }

// SystemPromptFromFile reads the system prompt from disk.
func (a *Agent) SystemPromptFromFile(p string) *Agent {
	data, err := os.ReadFile(p)
	if err != nil {
		a.cfg.SystemPrompt = ""
		return a.sync()
	}
	a.cfg.SystemPrompt = string(data)
	return a.sync()
}

// AllowedTools sets config.allowed_tools.
func (a *Agent) AllowedTools(t []string) *Agent { a.cfg.AllowedTools = t; return a.sync() }

// LLM sets config.llm (replaces v1 Model). Freeform preferred LLM id.
func (a *Agent) LLM(m string) *Agent { a.cfg.LLM = m; return a.sync() }

// HistoryLimit sets config.history_limit.
func (a *Agent) HistoryLimit(n int) *Agent { v := n; a.cfg.HistoryLimit = &v; return a.sync() }

// Componentable is any builder that can be embedded as an inline component
// in an Agent's components[]. Agent and Skill both implement it.
type Componentable interface {
	Build() (*Manifest, error)
}

// Component appends an inline sub-agent or sub-skill to components[].
// name is the local label; id is the canonical ns/name@version. child may be
// either *Agent or *Skill (anything implementing Componentable).
func (a *Agent) Component(name, id string, child Componentable) *Agent {
	a.sync()
	m, err := child.Build()
	if err != nil {
		// Defer error to Build() of this agent by appending nothing;
		// schema validation will catch any missing required component.
		return a
	}
	raw, _ := json.Marshal(m.Config)
	item := ComponentItem{
		Name:   name,
		ID:     id,
		Kind:   string(m.Kind),
		Config: raw,
	}
	if len(m.Files) > 0 {
		item.Files = m.Files
	}
	if m.Permissions != nil {
		item.Permissions = m.Permissions
	}
	if len(m.Dependencies) > 0 {
		item.Dependencies = m.Dependencies
	}
	if len(m.Components) > 0 {
		item.Components = m.Components
	}
	a.manifest.Components = append(a.manifest.Components, item)
	return a
}

// Ref appends an external ref component (any kind: tool, skill, or agent).
func (a *Agent) Ref(nsNameAtVersion string) *Agent {
	a.sync()
	a.manifest.Components = append(a.manifest.Components, ComponentItem{Ref: nsNameAtVersion})
	return a
}

// Flatten sets install.flatten merge rules.
func (a *Agent) Flatten(f InstallFlatten) *Agent {
	a.sync()
	if a.manifest.Install == nil {
		a.manifest.Install = &InstallBlock{}
	}
	a.manifest.Install.Flatten = &f
	return a
}

// Build returns the validated manifest.
func (a *Agent) Build() (*Manifest, error) { a.sync(); return a.base.build() }

// Pack writes a .aagent to outPath.
func (a *Agent) Pack(outPath string, opts ...PackOpt) (*Manifest, error) {
	a.sync()
	return a.base.packInternal(outPath, opts)
}

// Skill is a fluent builder for kind=skill packages.
type Skill struct {
	base
	cfg SkillConfig
}

// NewSkill constructs a Skill builder.
func NewSkill(name, version string) *Skill {
	s := &Skill{cfg: SkillConfig{Kind: "skill"}}
	s.base = newBase(name, version, KindSkill, s.cfg)
	return s
}

func (s *Skill) sync() *Skill { s.setConfig(s.cfg); return s }

// Description sets the human-readable description.
func (s *Skill) Description(d string) *Skill { s.manifest.Description = d; return s }

// Author sets the author field.
func (s *Skill) Author(a string) *Skill { s.manifest.Author = a; return s }

// License sets the license field.
func (s *Skill) License(l string) *Skill { s.manifest.License = l; return s }

// RequiresAlexandria sets the minimum Alexandria version.
func (s *Skill) RequiresAlexandria(v string) *Skill { s.manifest.RequiresAlexandria = v; return s }

// Dependency appends a dependency.
func (s *Skill) Dependency(d Dependency) *Skill {
	s.manifest.Dependencies = append(s.manifest.Dependencies, d)
	return s
}

// Dependencies replaces the dependencies slice.
func (s *Skill) Dependencies(ds []Dependency) *Skill { s.manifest.Dependencies = ds; return s }

// File appends a single file entry.
func (s *Skill) File(f FileEntry) *Skill { s.manifest.Files = append(s.manifest.Files, f); return s }

// Files replaces the files slice.
func (s *Skill) Files(fs []FileEntry) *Skill { s.manifest.Files = fs; return s }

// StageFile stages a file from disk and appends a matching files[] entry.
func (s *Skill) StageFile(srcPath string, entry FileEntry) *Skill {
	abs, _ := filepath.Abs(srcPath)
	s.staged = append(s.staged, stagedFile{archivePath: entry.ArchivePath, srcAbs: abs})
	return s.File(entry)
}

// ProvidesTools sets permissions.provides_tools.
func (s *Skill) ProvidesTools(t []string) *Skill { s.ensurePerms().ProvidesTools = t; return s }

// NeedsTools sets permissions.needs_tools.
func (s *Skill) NeedsTools(t []string) *Skill { s.ensurePerms().NeedsTools = t; return s }

// SuggestedRole sets permissions.suggested_role.
func (s *Skill) SuggestedRole(r string) *Skill { s.ensurePerms().SuggestedRole = r; return s }

// SystemPrompt sets config.system_prompt.
func (s *Skill) SystemPrompt(p string) *Skill { s.cfg.SystemPrompt = p; return s.sync() }

// AllowedTools sets config.allowed_tools.
func (s *Skill) AllowedTools(t []string) *Skill { s.cfg.AllowedTools = t; return s.sync() }

// LLM sets config.llm (replaces v1 ModelHint). Freeform preferred LLM id.
func (s *Skill) LLM(m string) *Skill { s.cfg.LLM = m; return s.sync() }

// Tags sets config.tags.
func (s *Skill) Tags(t []string) *Skill { s.cfg.Tags = t; return s.sync() }

// Build returns the validated manifest.
func (s *Skill) Build() (*Manifest, error) { s.sync(); return s.base.build() }

// Pack writes a .atool to outPath.
func (s *Skill) Pack(outPath string, opts ...PackOpt) (*Manifest, error) {
	s.sync()
	return s.base.packInternal(outPath, opts)
}

func shortName(n string) string {
	if i := strings.LastIndex(n, "/"); i >= 0 {
		return n[i+1:]
	}
	return n
}

File: packages/typescript/src/builders.ts
import { writeFileSync, mkdtempSync, mkdirSync, copyFileSync, readFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { dirname, join, resolve } from "node:path";
import { pack } from "./pack.js";
import { assertValid } from "./schema.js";
import type {
  AagentConfig, AtoolConfig, BundleConfig, ComponentItem, CredentialDecl, Dependency, EnvDecl, FileEntry,
  InlineComponent, InstallFlatten, K8sResources, LockEntry, Manifest, McpConfig, PackageDep,
  Permissions,
} from "./types.js";

abstract class Base<TConfig extends Manifest["config"]> {
  protected manifest: Manifest;

  constructor(name: string, version: string, kind: Manifest["kind"], config: TConfig) {
    this.manifest = {
      schema_version: "2",
      name,
      version,
      kind,
      description: "",
      config: config as Manifest["config"],
    };
  }

  description(d: string): this { this.manifest.description = d; return this; }
  author(a: string): this { this.manifest.author = a; return this; }
  license(l: string): this { this.manifest.license = l; return this; }
  requiresAlexandria(v: string): this { this.manifest.requires_alexandria = v; return this; }

  dependency(d: Dependency): this {
    (this.manifest.dependencies ??= []).push(d);
    return this;
  }
  dependencies(ds: Dependency[]): this { this.manifest.dependencies = ds; return this; }

  file(f: FileEntry): this { (this.manifest.files ??= []).push(f); return this; }
  files(fs: FileEntry[]): this { this.manifest.files = fs; return this; }

  protected ensurePerms(): Permissions {
    return (this.manifest.permissions ??= {});
  }
  providesTools(t: string[]): this { this.ensurePerms().provides_tools = t; return this; }
  needsTools(t: string[]): this { this.ensurePerms().needs_tools = t; return this; }
  suggestedRole(r: string): this { this.ensurePerms().suggested_role = r; return this; }

  build(): Manifest {
    return assertValid(JSON.parse(JSON.stringify(this.manifest)));
  }

  /**
   * Pack to a `.atool` / `.aagent`. By default, materialises a temp source dir
   * with `atool.json` and (for builders that staged file content) any staged
   * files; advanced callers can pass an existing source dir via `opts.srcDir`.
   */
  async pack(outPath: string, opts: { srcDir?: string } = {}): Promise<Manifest> {
    if (opts.srcDir) {
      const m = this.build();
      writeFileSync(join(opts.srcDir, "atool.json"), JSON.stringify(m, null, 2) + "\n");
      return pack(opts.srcDir, outPath);
    }
    const dir = mkdtempSync(join(tmpdir(), "alex-sdk-"));
    const m = this.build();
    writeFileSync(join(dir, "atool.json"), JSON.stringify(m, null, 2) + "\n");
    for (const [archivePath, srcAbs] of this.staged) {
      const dest = join(dir, archivePath);
      mkdirSync(dirname(dest), { recursive: true });
      copyFileSync(srcAbs, dest);
    }
    return pack(dir, outPath);
  }

  protected staged: Array<[string, string]> = [];
  /**
   * Stage a file from disk so `.pack()` can include it without a pre-laid-out
   * source dir. Automatically appends a matching `files[]` entry.
   */
  stageFile(srcPath: string, entry: Omit<FileEntry, "sha256">): this {
    this.staged.push([entry.archive_path, resolve(srcPath)]);
    this.file(entry);
    return this;
  }
}

/**
 * Binary tool builder. Emits `kind = atool` (native gRPC `ToolService`) by
 * default; calling `.transport("http" | "sse")` re-taxes the package to
 * `kind = mcp` (MCP JSON-RPC/SSE). `.transport("grpc")` keeps it an atool.
 * This mirrors EE: a "tool" is either an mcp daemon or an atool daemon,
 * discriminated by the wire protocol the orchestrator speaks.
 */
export class Tool extends Base<McpConfig | AtoolConfig> {
  declare protected manifest: Manifest & { config: McpConfig | AtoolConfig };
  constructor(name: string, version: string) {
    // Default transport is gRPC => atool, matching EE's AtoolConfig default.
    super(name, version, "atool", { kind: "atool", binary: "" });
  }
  binary(p: string): this { this.manifest.config.binary = p; return this; }
  port(p: number): this { this.manifest.config.default_port = p; return this; }
  /**
   * Pick the wire protocol — and thereby the package kind:
   *   "grpc"       => kind atool (native ToolService)
   *   "http"|"sse" => kind mcp   (MCP JSON-RPC/SSE)
   */
  transport(t: "grpc" | "http" | "sse"): this {
    const kind = t === "grpc" ? "atool" : "mcp";
    this.manifest.kind = kind;
    this.manifest.config.kind = kind;
    this.manifest.config.transport = t;
    return this;
  }
  args(a: string[]): this { this.manifest.config.args = a; return this; }
  /** Contract/ABI major this tool exposes over its wire protocol (EE default 1). */
  interfaceMajor(n: number): this { this.manifest.config.interface_major = n; return this; }

  /**
   * Declare this as a **code-less** tool that binds to a native orchestrator
   * handler INSTEAD of shipping a binary (closed set, currently
   * `"emit_trigger"`). Drops the default-seeded empty `binary`. A code-less tool
   * must also declare its {@link Tool.inputSchema} — there is no daemon to
   * advertise it.
   */
  nativeHandler(name: string): this {
    const cfg = this.manifest.config as AtoolConfig;
    cfg.native_handler = name;
    delete cfg.binary;
    return this;
  }

  /**
   * Declare the tool's full input contract as an embedded JSON Schema object.
   * Required for a code-less tool ({@link Tool.nativeHandler}); optional static
   * fallback for a coded tool.
   */
  inputSchema(schema: Record<string, unknown>): this {
    (this.manifest.config as AtoolConfig).input_schema = schema;
    return this;
  }

  /**
   * Declare a secret credential this tool reads from an environment variable.
   * No secret *value* is ever placed in the package — the operator binds the
   * value into the deployment-shape secret backend at install time. `secret`
   * defaults `true`, `rotation` defaults `"respawn"`.
   */
  credential(
    env: string,
    opts: { required?: boolean; secret?: boolean; description?: string; rotation?: "respawn" | "oauth-refresh" } = {},
  ): this {
    const c: CredentialDecl = {
      env,
      secret: opts.secret ?? true,
      required: opts.required ?? false,
      rotation: opts.rotation ?? "respawn",
    };
    if (opts.description) c.description = opts.description;
    (this.manifest.config.credentials ??= []).push(c);
    return this;
  }

  /**
   * Declare a non-secret config environment variable this tool reads, with an
   * optional literal `default`. Values are stored inline by the operator, not as
   * a secret ref.
   */
  env(name: string, opts: { default?: string; required?: boolean } = {}): this {
    const e: EnvDecl = { name };
    if (opts.default !== undefined) e.default = opts.default;
    if (opts.required) e.required = true;
    (this.manifest.config.env ??= []).push(e);
    return this;
  }
  k8sImage(img: string): this { this.manifest.config.k8s_image = img; return this; }
  k8sCapabilities(c: string[]): this { this.manifest.config.k8s_capabilities = c; return this; }
  k8sPort(p: number): this { this.manifest.config.k8s_port = p; return this; }
  k8sTransport(t: "grpc" | "http" | "sse"): this { this.manifest.config.k8s_transport = t; return this; }
  k8sResources(r: K8sResources): this { this.manifest.config.k8s_resources = r; return this; }
  k8sMinWarm(n: number): this { this.manifest.config.k8s_min_warm = n; return this; }
  k8sIdleTimeout(seconds: number): this { this.manifest.config.k8s_idle_timeout_seconds = seconds; return this; }
}

export class Agent extends Base<AagentConfig> {
  declare protected manifest: Manifest & { config: AagentConfig };
  constructor(name: string, version: string) {
    super(name, version, "aagent", { kind: "aagent", system_prompt: "" });
  }
  systemPrompt(s: string): this { this.manifest.config.system_prompt = s; return this; }
  systemPromptFromFile(p: string): this {
    this.manifest.config.system_prompt = readFileSync(p, "utf8");
    return this;
  }
  allowedTools(t: string[]): this { this.manifest.config.allowed_tools = t; return this; }
  /** Preferred model backend id (EE `config.model`). Replaces v1 `.llm()`. */
  model(id: string): this { this.manifest.config.model = id; return this; }
  historyLimit(n: number): this { this.manifest.config.history_limit = n; return this; }
  /** Prompt composition mode against `extends` bases: "append" (default) | "replace". */
  promptMode(m: "append" | "replace"): this { this.manifest.config.prompt_mode = m; return this; }

  /**
   * Append a base package this aagent extends. aagent-only in EE — the flatten
   * resolver composes base→child. Rejected by validation on mcp/atool kinds.
   */
  extend(base: PackageDep): this {
    (this.manifest.extends ??= []).push(base);
    return this;
  }
  extendsPackages(bases: PackageDep[]): this { this.manifest.extends = bases; return this; }

  /** Append a resolved inheritance lockfile entry (aagent-only). */
  lock(entry: LockEntry): this {
    (this.manifest.lockfile ??= []).push(entry);
    return this;
  }
  lockfile(entries: LockEntry[]): this { this.manifest.lockfile = entries; return this; }

  /**
   * Append an inline sub-agent component.
   * `name` is the local label; `id` is the canonical ns/name@version.
   * Both Agent and Skill children emit kind=aagent.
   */
  component(name: string, id: string, child: Agent | Skill): this {
    const childManifest = child.build();
    const item: InlineComponent = {
      name,
      id,
      kind: "aagent",
      config: childManifest.config as AagentConfig,
    };
    if (childManifest.files) item.files = childManifest.files;
    if (childManifest.permissions) item.permissions = childManifest.permissions;
    if (childManifest.dependencies) item.dependencies = childManifest.dependencies;
    if (childManifest.components) item.components = childManifest.components;
    (this.manifest.components ??= []).push(item);
    return this;
  }

  /**
   * Append an external ref component (any kind: mcp/atool tool, or aagent).
   * Binary tools may ONLY appear as refs (never inline) per the schema.
   */
  ref(nsNameAtVersion: string): this {
    (this.manifest.components ??= []).push({ ref: nsNameAtVersion });
    return this;
  }

  /**
   * Set install.flatten merge rules (only meaningful on aagents with components[]).
   */
  flatten(rules: InstallFlatten): this {
    (this.manifest.install ??= {}).flatten = rules;
    return this;
  }
}

/**
 * Skill builder. A "skill" is reusable prompt text — EE has no standalone skill
 * kind, so this emits `kind = aagent` whose only content is `system_prompt`.
 */
export class Skill extends Base<AagentConfig> {
  declare protected manifest: Manifest & { config: AagentConfig };
  constructor(name: string, version: string) {
    super(name, version, "aagent", { kind: "aagent", system_prompt: "" });
  }
  systemPrompt(s: string): this { this.manifest.config.system_prompt = s; return this; }
  systemPromptFromFile(p: string): this {
    this.manifest.config.system_prompt = readFileSync(p, "utf8");
    return this;
  }
  allowedTools(t: string[]): this { this.manifest.config.allowed_tools = t; return this; }
  /** Preferred model backend id (EE `config.model`). Replaces v1 `.modelHint()`. */
  model(id: string): this { this.manifest.config.model = id; return this; }
}

/**
 * Bundle builder. A bundle is a NON-callable named set of member tools — the
 * unit a "role" (doer/delegator/file-handler) is made of. It ships no binary,
 * no native_handler, no input_schema, no model, no system_prompt; it is pure
 * composition. Its doctrine/"skill" (the stance) lives in the top-level
 * `.description(...)`. Emits `kind = "bundle"` with `config.tools`.
 */
export class Bundle extends Base<BundleConfig> {
  declare protected manifest: Manifest & { config: BundleConfig };
  constructor(name: string, version: string) {
    super(name, version, "bundle", { kind: "bundle", tools: [] });
  }
  /** Add one member tool reference (optionally `name@major`). */
  tool(ref: string): this { this.manifest.config.tools.push(ref); return this; }
  /** Replace the member tool list. At least one is required by the schema. */
  tools(refs: string[]): this { this.manifest.config.tools = [...refs]; return this; }
}
